uportal-war/src/main/java/org/jasig/portal/groups/pags/testers/AdHocGroupTester.java

```diff
@@ -0,0 +1,172 @@
+                groups.add(entityGroup);
+            } else {
+                if (throwOnFail) {
+                    logger.error("Could not find group named {}", groupName);
+                    throw new IllegalArgumentException("Could not find group named " + groupName);
+                } else {
+                    logger.warn("Could not find group named {}", groupName);
+                }
+            }
+        }
+    }
+
+    /**
+     * Create a hashcode based on the includes/excludes groups.
+     *
+     * @return      hashcode for this instance
+     */
+    private String calcHashcode() {
+        StringBuilder hash = new StringBuilder("__");
+        for (IEntityGroup group : includes) {
+            hash.append(group.getKey());
+        }
+        hash.append("^");
+        for (IEntityGroup group : excludes) {
+            hash.append(group.getKey());
+        }
+        hash.append("_#");
+        return hash.toString();
+    }
+
+    @Override
+    public boolean test(IPerson person) {
+        String personHash = person.getEntityIdentifier().getKey() + hashcode + Thread.currentThread().getId();
+        logger.debug(personHash);
+        if (currentPersons.contains(personHash))
+            return false; // stop recursing
+        currentPersons.add(personHash);
+
+        IGroupMember gmPerson = findPersonAsGroupMember(person);
+        boolean pass = true;
+
+        Iterator<IEntityGroup> groups = includes.iterator();
+        while (pass && groups.hasNext()) {
+            IEntityGroup group = groups.next();
+            if (!gmPerson.isDeepMemberOf(group)) {
+                logger.debug("!! {} is not a member of {}", person.getUserName(), group.getDescription());
+                pass = false;
+            } else {
+                logger.debug("{} is a member of {}", person.getUserName(), group.getDescription());
+            }
+        }
+        groups = excludes.iterator();
+        while (pass && groups.hasNext()) {
+            IEntityGroup group = groups.next();
+            if (gmPerson.isDeepMemberOf(group)) {
+                logger.debug("!! {} is a member of {}", person.getUserName(), group.getDescription());
+                pass = false;
+            } else {
+                logger.debug("{} is a not member of {}", person.getUserName(), group.getDescription());
+            }
+        }
+
+        currentPersons.remove(personHash);
+        return pass;
+    }
+
+    /**
+     * Find {@link IEntityGroup} from group name.
+     *
+     * @param groupName     name of group to search from {@code GroupService}
+     * @return              {@code IEntityGroup} with given name or null if no group with given name found
+     * @see                 org.jasig.portal.services.GroupService
+     */
+    private IEntityGroup findGroupByName(String groupName) {
+        EntityIdentifier[] identifiers = GroupService.searchForGroups(groupName, GroupService.IS, EntityEnum.GROUP.getClazz());
```

Comment:
I suspect the 3rd argument should be `IPerson.class` -- or, alternately `EntityEnum.PERSON`

I'd rather not refer here to something (`EntityEnum`) in a package called `org.jasig.portal.portlets.groupselector` (to me it feels like coloring-out-of-the-lines) but I see a lot of code all over that does that already.  (Perhaps the enum should be moved.)